### ISSUE TITLE
derp: occasionally reconnect to primary node within a region

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#12724]
Each tailnet gets a different deterministically random ordering of DERP nodes in a region.

So generally all nodes in a tailnet are connected to the same of ~3 DERP nodes within a region.

But sometimes the main choice fails to connect for whatever reasons and clients get split between nodes and the derp server forwards packets ("mesh") between itself (within a region, only).

But that meshing isn't super optimized. We should do that, but this bug is about something else @raggi mentioned wanting to do: make the client itself occasionally try to get back on to the primary node if it finds itself on the 2nd+.

The simple thing to do would be to just break the connection every 30 minutes and retry and hope for the best. That involves a blip of data plane downtime, though. And there's no reason to do that in the client, since we could just as well do that server side from derper. It also might not even work: it might be that one derp node's IP is blocked in a region but another one is not. We've oddly seen that before.

So the fancier thing to do in the client is try to connect to the primary node in the background while keeping the non-ideal one active until the last second, and then ~atomically (as much as possible) switch them over, only once we see the TCP+TLS connects work.

The code doesn't make this trivial, but filing a tracking bug.

**Comment by phinze** [comment]
@mpminardi is going to take a look at this this week

**Comment by mpminardi** [comment]
Putting the work and design for this solution on hold for now. The current plan is to utilize `frameClosePeer` after DERP server deploys to rebalance the clients on given servers / force clients back onto their ideal DERP nodes in a region as this may get us most of what we want with a shorter path towards value being delivered.

This issue will remain open for tracking the design / work for future client side logic to aid with this balancing if this is till determined to be necessary after the above is done.

**Comment by mikeodr** [comment]
buildkite deploys now prune automatically after a regional deploy.

Moving back to no-status as a standin ticket for the routing/balancing work to be done.

**Comment by mikeodr** [comment]
https://github.com/tailscale/corp/issues/27614 Is a dupe but has additional info.
